<h1 style=color:blue;>Booststrapping</h1>

Bootstrap is a powerful, computer-based method for statistical inference without relying on too many assumption. in fact, it is widely applied in other statistical inference such as confidence interval, regression model, even the field of machine learning.
This markdown notebook focus on introducing the core concepts of Bootstrap than its application.

<h1 style=color:blue;>The Boostrapping Sampling Method</h1>

he basic idea of bootstrap is make inference about a estimatelike  sample mean for a population parameter θ like population mean on sample data.It is a resampling method by independently sampling with replacement from an existing sample data with same sample size n, and performing inference among these resampled data.

Generally, bootstrap involves the following steps:

<img src="bootstrapping.png" alt="Generally, bootstrap involves the following steps:">

1.A sample from population with sample size n.


2.Draw a sample from the original sample data with replacement with size n, and replicate B times, each re-sampled sample is called a Bootstrap Sample, and there will totally B Bootstrap Samples.

3.Evaluate the statistic of θ for each Bootstrap Sample, and there will be totally B estimates of θ.

4.Construct a sampling distribution with these B Bootstrap statistics and use it to make further statistical inference, such as:

* Estimating the standard error of statistic for θ.
* Obtaining a Confidence Interval for θ.

### How and why does bootstrap work?

This lecture  ,I will divide  into three parts?

* initial motivation
* simulation technique or  estimated variance of statistic by resampling?
* Idea behind draw a sample from the original sample with replacemen

### 1. Initial Motivation- The Estimator’s Standard Error

The core idea of bootstrap technique is for making certain kinds of statistical inference with the help of modern computer power. When Efron introduced the method, it was particularly motivated by evaluating of the accuracy of an estimator in the field of statistic inference. Usually, estimated standard error are an first step toward thinking critically about the accuracy statistical estimates.

how bootstrap works and how an estimator’s standard error plays an important role, let’s start with a simple case.

#### Scenario Case

 *Imagine that you want to summarize how many times a day do students pick up their smartphone in  lab with totally 100 students. It's hard to summarize the number of pickups in whole lab like a census way. Instead, we make an online survey which also provided the pickup-counting APP. In the next few days, we receive 30 students responses with their number of pickups in a given day. We calculated the mean of these 30 pickups and got an estimate for pickups is 228.06 times.*


In [1]:
import numpy as np
import matplotlib.pyplot as plt

In [9]:
# construct a population pickups for our lab
np.random.seed(42)
pickups = np.random.randint(0,500 , size=100)
pickups

array([102, 435, 348, 270, 106,  71, 188,  20, 102, 121, 466, 214, 330,
       458,  87, 372,  99, 359, 151, 130, 149, 308, 257, 343, 491, 413,
       293, 385, 191, 443, 276, 160, 459, 313,  21, 252, 235, 344,  48,
       474,  58, 169, 475, 187, 463, 270, 189, 445, 174, 445,  50, 363,
        54, 243, 319, 130, 484, 306, 134,  20, 328, 166, 273, 387,  88,
       315,  13, 241, 264, 345,  52, 385, 339,  91, 366, 443, 454, 427,
       263, 430,  34, 205,  80, 419,  49, 359, 387,   1, 389,  53, 105,
       259, 309, 476, 190, 401, 217,  43, 161, 201])

In [10]:
# population mean
pickups.mean()

252.7

In [11]:
# population standard deviation
pickups.std()

144.25342283634035

In [12]:
# draw a sample from population
sample = np.random.choice(pickups, size=30)
sample

array([166, 201, 458, 190, 445,  87, 385, 427, 387, 166, 474,  49, 430,
       205,  54, 343, 413, 389,  20,  58, 191,  87, 463,  88, 389,  52,
       102,   1, 102,  20])

In [13]:
# our first sample mean
sample_mean = sample.mean()
sample_mean

228.06666666666666

In [14]:
# standard deiveation for this sample
sample_std = np.std(sample, ddof=1)
sample_std

166.96890756052164

In [15]:
# estimated standard error for the sapmle mann
sample_std/(30 ** 0.5)

30.48421235763086

The process above is called a **point estimate**. What we would like to know is the true number of pickups in whole lab. We don’t have census data, what we can do is just evaluate the **population parameter** through an **estimator** based on an observed sample, and then get an **estimate** as the evaluation of average smartphone usage in the lab.

* **Estimator/Statistic:** A rule for calculating an estimate. In this case is Sample mean, always denoted as X̄.

* **Population Parameter:** Numeric summary about a population. In this case is the average time of phone pickups per day in our lab, always denoted as μ.

### The Standard Error

The standard error of an estimator is it’s standard deviation. It tells us how far your sample estimate deviates from the actual parameter. If the standard error itself involves unknown parameters, we used the estimated standard error by replacing the unknown parameters with an estimate of the parameters.

Let’s take an example. In our case, our estimator is sample mean, and for sample mean(and nearly only one!), we have an simple formula to easily obtain it’s standard error.

<img src="stdv.png" alt="Generally, bootstrap involves the following steps:">

However, the standard deviation of population σ is always unknown in real world, so the most common measurement is the estimated standard error, which use the sample standard deviation S as a estimated standard deviation of the population:

<img src="samplestdv.png" alt="Generally, bootstrap involves the following steps:">

In our case, we have sample with 30, and sample mean is 228.06, and the sample standard deviation is 166.97, so our estimated standard error for our sample mean is 166.97/ √30 = 30.48.